File: Android/app/src/main/java/knk/ee/neverland/fragments/FeedFragment.java
package knk.ee.neverland.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;

import knk.ee.neverland.R;
import knk.ee.neverland.feed.FeedElementAdapter;
import knk.ee.neverland.feed.FeedElement;

public class FeedFragment extends Fragment {
    public FeedFragment() {
        // Required empty public constructor
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initializeFeedListView();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_feed, container, false);
    }

    private void initializeFeedListView() {
        ListView feedListView = getView().findViewById(R.id.feed_listview);
        feedListView.setAdapter(new FeedElementAdapter(getContext(), createDummyList()));
    }

    private List<FeedElement> createDummyList() {
        List<FeedElement> list = new ArrayList<>();

        for (int i = 0; i < 50; i++) {
            list.add(new FeedElement());
        }

        return list;
    }
}

File: Android/app/src/main/java/knk/ee/neverland/activities/fragments/FeedFragment.kt
package knk.ee.neverland.activities.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ListView
import knk.ee.neverland.R
import knk.ee.neverland.api.DefaultAPI
import knk.ee.neverland.api.FeedScope
import knk.ee.neverland.models.Proof
import knk.ee.neverland.utils.APIAsyncRequest
import knk.ee.neverland.views.feedview.FeedElementAdapter

class FeedFragment : Fragment() {
    private lateinit var feedElementAdapter: FeedElementAdapter
    private lateinit var feedListSwipeLayout: SwipeRefreshLayout

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeFeedListView()

        feedListSwipeLayout = view.findViewById(R.id.feed_list_swipe_layout)
        feedListSwipeLayout.setOnRefreshListener({
            runGetProofsTask(true)
        })

        view.findViewById<Toolbar>(R.id.toolbar).title = getString(R.string.feed_fragment_title)

        runGetProofsTask(false)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_feed, container, false)
    }

    private fun runGetProofsTask(updating: Boolean) {
        APIAsyncRequest.Builder<List<Proof>>()
            .before { feedListSwipeLayout.isRefreshing = true && updating }
            .request { DefaultAPI.proofAPI.getProofs(FeedScope.WORLD) }
            .handleResult { result -> feedElementAdapter.updateList(result!!) }
            .setContext(view!!.context)
            .showMessages(true)
            .after { feedListSwipeLayout.isRefreshing = false }
            .onAPIFailMessage { R.string.error_getting_proofs_failed }
            .finish()
            .execute()
    }

    private fun initializeFeedListView() {
        val feedListView = view!!.findViewById<ListView>(R.id.feed_listview)
        feedElementAdapter = FeedElementAdapter(context!!)
        feedListView.adapter = feedElementAdapter
        // feedListView.emptyView = // TODO: !!!
    }
}
